### PR TITLE
[DNM] Fixes bug 1342081 - Upgraded ES libraries and related code.

### PR DESCRIPTION
**Pull request by adngdb**
This commit upgrades our ES modules, ``elasticsearch`` and ``elasticsearch-dsl``, as well as all code related to them. It makes sure that unit tests pass using ES 5.1 as the database.

### CODE REVIEW

File: scripts/start-travis-elasticsearch.sh [hunk 1]
```diff
@@ -4,12 +4,9 @@ set -ex
...
-tar zxf elasticsearch-1.4.5.tar.gz
-echo "script.disable_dynamic: false" > elasticsearch.yml
-echo "index.number_of_shards: 1" >> elasticsearch.yml
-echo "index.number_of_replicas: 0" >> elasticsearch.yml
-# the -d flag starts it as a daemon in the background
-./elasticsearch-1.4.5/bin/elasticsearch -d --config=elasticsearch.yml
+wget --no-check-certificate 'https://artifacts.elastic.co/downloads/elasticsearch/elasticsearch-5.1.1.deb'
+sudo dpkg -i --force-confnew elasticsearch-5.1.1.deb
+
+sudo service elasticsearch start
```
Comment (willkg) [inline scripts/start-travis-elasticsearch.sh]: Do we want to stomp on values in the yml file at all before running tests?

File: socorro/external/es/connection_context.py [hunk 2]
```diff
@@ -129,6 +130,12 @@ def indices_client(self, name=None):
             elasticsearch.client.IndicesClient(self.connection())
         )
 
+    def cat_client(self):
```
Comment (willkg) [inline socorro/external/es/connection_context.py]: What does this do? Can you add a docstring?

The elasticsearch-py docs aren't helpful at all:

https://elasticsearch-py.readthedocs.io/en/master/api.html?highlight=catclient#cat

File: socorro/unittest/external/es/base.py [hunk 3]
```diff
@@ -172,12 +170,17 @@
         'form_field_choices': None,
         'permissions_needed': [],
         'default_value': None,
-        'has_full_version': False,
+        'has_full_version': True,
```
Comment (willkg) [inline socorro/unittest/external/es/base.py]: Is this a memory address? If so, why would we want to store it as `text`?

File: socorro/unittest/external/es/base.py [hunk 4]
```diff
@@ -135,12 +128,17 @@
...
         'is_exposed': True,
         'is_returned': True,
         'is_mandatory': False,
         'storage_mapping': {
-            'type': 'string'
+            'type': 'text',
```
Comment (willkg) [inline socorro/unittest/external/es/base.py]: Why would we want to store the `release_channel` as text?

File: socorro/unittest/external/es/base.py [hunk 5]
```diff
@@ -284,7 +289,12 @@
         'permissions_needed': [],
         'query_type': 'enum',
         'storage_mapping': {
-            'type': 'string'
+            'type': 'text',
```
Comment (willkg) [inline socorro/unittest/external/es/base.py]: Why would we want to store this as text?

File: socorro/unittest/external/es/base.py [hunk 6]
```diff
@@ -301,11 +311,10 @@
         'permissions_needed': [],
         'query_type': 'string',
         'storage_mapping': {
-            'type': 'string',
+            'type': 'text',
             'fields': {
                 'full': {
-                    'type': 'string',
-                    'index': 'not_analyzed',
+                    'type': 'keyword',
```
Comment (willkg) [inline socorro/unittest/external/es/base.py]: Why would we want to store this as keyword?

File: socorro/unittest/external/es/base.py [hunk 7]
```diff
@@ -377,7 +384,12 @@
         'permissions_needed': [],
         'query_type': 'enum',
         'storage_mapping': {
-            'type': 'string'
+            'type': 'text',
```
Comment (willkg) [inline socorro/unittest/external/es/base.py]: Why would we want to store this as text?

File: socorro/unittest/external/es/base.py [hunk 8]
```diff
@@ -463,11 +474,10 @@
         'permissions_needed': [],
         'query_type': 'enum',
         'storage_mapping': {
-            'type': 'string',
+            'type': 'text',
```
Comment (willkg) [inline socorro/unittest/external/es/base.py]: Why would we want to store this as text?

File: socorro/unittest/external/es/base.py [hunk 9]
```diff
@@ -509,11 +518,10 @@
...
```
Comment (willkg) [inline socorro/unittest/external/es/base.py]: Why would we want to store this as text?

File: socorro/unittest/external/es/base.py [hunk 10]
```diff
@@ -654,14 +664,19 @@
         ],
         'query_type': 'enum',
         'storage_mapping': {
-            'type': 'string'
+            'type': 'text',
```
Comment (willkg) [inline socorro/unittest/external/es/base.py]: Why would we want to store this as text?

File: socorro/unittest/external/es/test_supersearch.py [hunk 11]
```diff
@@ -1569,33 +1550,33 @@ def test_get_with_number_histogram(self):
...
                             'term': 'Linux',
                             'count': 1
                         }
                     ],
+                    'product': [
+                        {
+                            'term': 'NightTrain',
+                            'count': 1
+                        },
+                    ],
```
Comment (willkg) [inline socorro/unittest/external/es/test_supersearch.py]: Why did this move? Is it a cosmetic change?

File: scripts/start-travis-elasticsearch.sh (hunk 1 above)
Comment (adngdb) [inline scripts/start-travis-elasticsearch.sh]: I think we're fine using the default values. Anyway, we can't set the config options we use to anymore, because ES won't accept those values in the yml file.

File: socorro/external/es/mappings/correlations_index_settings.json [hunk 12]
```diff
@@ -4,9 +4,6 @@
             "_all": {
                 "enabled": false
             },
-            "_source": {
-                "compress": true
-            },
```
Comment (adngdb) [inline socorro/external/es/mappings/correlations_index_settings.json]: Yes. It's not just that we can, it's that the option is not accepted anymore, so we have to remove it. :)

File: socorro/external/es/new_crash_source.py [hunk 13]
```diff
@@ -53,34 +54,30 @@ def new_crashes(self, date, product, versions):
...
+        filter_args = {
+            'processed_crash.date_processed': {
+                'gte': date.isoformat(),
+                'lt': next_day.isoformat(),
+            },
+        }
+        search = search.filter('range', **filter_args)
+
+        # Turn that search body into a dict that we can send to Elasticsearch.
+        query = search.to_dict()
```
Comment (adngdb) [inline socorro/external/es/new_crash_source.py]: Yes, by default the library applies an ``AND`` boolean operator between filters. So all of the filters above must be true in the query.

File: socorro/external/es/new_crash_source.py [hunk 14]
```diff
@@ -92,11 +89,11 @@ def new_crashes(self, date, product, versions):
                 scroll='2m',  # keep the "scroll" connection open for 2 minutes
                 index=es_index,
                 doc_type=es_doctype,
-                fields=['crash_id'],
                 query=query,
+                _source=False,
             )
             for hit in res:
-                crash_ids.append(hit['fields']['crash_id'][0])
+                crash_ids.append(hit['_id'])
```
Comment (adngdb) [inline socorro/external/es/new_crash_source.py]: It's always been the case, and I used this opportunity to make this a little more efficient by not needing any field to be returned.

File: socorro/middleware/search_common.py [hunk 15]
```diff
@@ -135,7 +135,7 @@ def build_filters(self, fields):
                 ))
 
                 # Add an interval field.
-                default_interval = 1
+                default_interval = 50
```
Comment (adngdb) [inline socorro/middleware/search_common.py]: That's the Elasticsearch default now. This change is a security against the "matrix" problem I had while testing. If we set it to 1, then any histogram over a field that contains big numbers will return a huge number of results and will cause out-of-memory problems. By increasing the default interval we can mitigate that problem and ensure users won't accidentally hit it.

File: socorro/unittest/external/es/test_supersearch.py (hunk 11 above)
Comment (adngdb) [inline socorro/unittest/external/es/test_supersearch.py]: No, that's just because I was stupid when debugging a test failure and thought it was an array and not a dict. :)

File: socorro/external/es/supersearch.py [hunk 16]
```diff
@@ -99,19 +99,20 @@ def format_field_names(self, hit):
...
-                    hit[field] = hit[field][0]
+        def flatten(d, parent_key='', sep='.'):
+            items = []
+            for k, v in d.items():
+                new_key = parent_key + sep + k if parent_key else k
+                if isinstance(v, collections.MutableMapping):
+                    items.extend(flatten(v, new_key, sep=sep).items())
+                else:
+                    items.append((new_key, v))
+            return dict(items)
```
Comment (adngdb) [inline socorro/external/es/supersearch.py]: You're right, so I started moving it to ``socorro/lib/util.py``, and I feel like I've open Pandora's box... That file is a huge mess, I'm trying to clean it up, that will add a bunch of code changes to this PR but it will be in a separate commit.

File: socorro/unittest/external/es/base.py (hunk 6 above)
Comment (adngdb) [inline socorro/unittest/external/es/base.py]: This multi-fields storage serves a simple purpose: when searching, we want to have access to the content separated as terms, and when aggregating, we want to see it as a single block. In this case, take a signature like ``oom | ns::something<T> | randomFunc``. We want it to be stored as ``[oom, ns, something, randomfunc]`` to be able to retrieve it when doing ``<signature> <has terms> <oom>`` (that's why we store it as ``text``) but when showing a list of top signatures, we want to see it as ``oom | ns::something<T> | randomFunc`` (that's why we have a second storage as a ``keyword``, under the ``.full`` key name). Super Search uses the ``has_full_version`` attribute to see if there's a ``.full`` sub-key, and use it in aggregation if so.

Comment (adngdb) [inline socorro/unittest/external/es/base.py]: Also, all of these are only ever used for unit tests. The actual mapping that we use in production are inside our database, in the Super Search Fields list. See https://crash-stats.mozilla.com/admin/supersearch-fields/

Comment (adngdb) [inline socorro/unittest/external/es/base.py]: However you were right that some of those choices I made where dumb, so I went through all of them and fixed a few. Now all of them should be coherent with the data we have in production.

Comment (willkg) [inline socorro/unittest/external/es/base.py]: Awesome--this makes total sense.

File: scripts/start-travis-elasticsearch.sh (hunk 1 above)
Comment (willkg) [inline scripts/start-travis-elasticsearch.sh]: Makes sense!

File: socorro/middleware/search_common.py (hunk 15 above)
Comment (willkg) [inline socorro/middleware/search_common.py]: Makes sense!

File: socorro/external/es/new_crash_source.py (hunk 14 above)
Comment (willkg) [inline socorro/external/es/new_crash_source.py]: Makes sense!

File: socorro/lib/util.py [hunk 17]
```diff
@@ -2,225 +2,229 @@
...
+                    "e": 3
+                }
+            }
+        }
+        OUT:
+        {
+            "a": 1,
+            "b.c": 2,
+            "b.d.e": 3
+        }
```
Comment (willkg) [inline socorro/lib/util.py]: I tend to do this either using code blocks or doctests. One advantage of doctests is that you can test them and thus know when your examples are wrong. Both code blocks and doctests will get syntax highlighting when rendered in Sphinx.

I have no idea if this is rendered in Sphinx or not. If it is, it's worth changing.

Something like this:

```
For example:

>>> flatten({'a': 1, 'b': {'c': 2, 'd': {'e': 3}}})
{'a': 1, 'b.c': 2, 'b.d.e': 3}

```

**Comment by willkg** [comment]
Everything here seems fine to me. r+

**Comment by adngdb** [comment]
Just found out that we don't have any unit tests for the ``missing`` feature of Super Search. I need to add that (and fix the feature). Also, the schema still needs to be tested and validated more. Right now basic functionalities work, but I need to do further testing.

**Comment by adngdb** [comment]
I have been testing locally with data from downloaded from prod, and with an ES 5.1 database and the latest version of the schema that is in this PR. Everything seems to be working fine now!

@willkg mind taking a look at the last commit I added? It turned out we were missing unit tests for the ``__null__`` operator.

File: scripts/data/supersearch_fields.json [hunk 18]
```diff
@@ -338,9 +336,7 @@
         "is_returned": false,
         "query_type": "enum",
         "permissions_needed": [],
-        "storage_mapping": {
-            "type": "string"
-        },
+        "storage_mapping": null,
```
Comment (willkg) [inline scripts/data/supersearch_fields.json]: What does it mean to change the `storage_mapping` to null?

File: scripts/data/supersearch_fields.json [hunk 19]
```diff
@@ -1207,26 +1195,6 @@
...
-        },
-        "namespace": "processed_crash",
-        "name": "dump",
-        "has_full_version": false,
-        "data_validation_type": "str",
-        "in_database_name": "dump",
-        "is_exposed": false,
-        "form_field_choices": [],
-        "description": ""
-    },
```
Comment (willkg) [inline scripts/data/supersearch_fields.json]: This dumps the `dump` field. Why? Does anything use the `dump` field?

File: scripts/data/supersearch_fields.json [hunk 20]
```diff
@@ -2865,29 +2811,27 @@
                     "dynamic": "true",
                     "properties": {
                         "classification_version": {
-                            "type": "string",
-                            "analyzer": "keyword"
+                            "type": "keyword"
                         },
                         "classification_data": {
-                            "type": "string"
+                            "type": "text"
```
Comment (willkg) [inline scripts/data/supersearch_fields.json]: What does the `classification_data` field hold?

File: scripts/data/supersearch_fields.json [hunk 21]
```diff
@@ -3450,7 +3386,7 @@
         "query_type": "enum",
         "permissions_needed": [],
         "storage_mapping": {
-            "type": "string"
+            "type": "integer"
```
Comment (willkg) [inline scripts/data/supersearch_fields.json]: This switched from `string` to `integer`. Does that sound right? Are all arm version values strictly integers?

File: socorro/external/es/supersearch.py [hunk 22]
```diff
@@ -324,7 +324,9 @@ def get(self, **kwargs):
                         operator_range[param.operator]: param.value
                     }
                 elif param.operator == '__null__':
-                    filter_type = 'missing'
+                    filter_type = 'exists'
+                    # Need to reverse the 'not' operator.
+                    param.operator_not = not param.operator_not
```
Comment (willkg) [inline socorro/external/es/supersearch.py]: Clever!

**Comment by willkg** [comment]
Ugh. The GH interface for reviews totally sucks if you're looking at specific commits. That's frustrating.

I have questions in the second-to-last commit.

File: scripts/data/supersearch_fields.json (hunk 18 above)
Comment (adngdb) [inline scripts/data/supersearch_fields.json]: That this field will not have a storage mapping in Elasticsearch. Basically the impact is very low, we will just rely on ES's default mechanism. I'm doing this here because that field is not exposed or returned in any way, it's an old field that we don't use anymore.

File: scripts/data/supersearch_fields.json (hunk 19 above)
Comment (adngdb) [inline scripts/data/supersearch_fields.json]: That field doesn't exist anymore. It was a bit stupid to remove it there, it's out of scope, sorry about that.

File: scripts/data/supersearch_fields.json (hunk 21 above)
Comment (adngdb) [inline scripts/data/supersearch_fields.json]: According to https://crash-stats.mozilla.com/search/?_facets=support_classification that field contains only integers.

File: scripts/data/supersearch_fields.json (hunk 20 above)
Comment (adngdb) [inline scripts/data/supersearch_fields.json]: All I remember is that it has text about the classification. But that is not exposed anywhere, so it doesn't really matter.

**Comment by willkg** [comment]
This looks good to me. r+

File: requirements.txt [hunk 23]
```diff
@@ -59,12 +59,12 @@ psycopg2==2.6.2 \
...
-elasticsearch==1.9.0 \
-    --hash=sha256:cbee23c29f9947b1b78be5c0f741c1f8b504d3c53163b1355ea08f69cbd0fc0c \
-    --hash=sha256:eb0d4b5b20a4ccf9e49bd20498b6215d5e960d8a63a0eae44cd4431e3f303a24
-elasticsearch-dsl==0.0.11 \
-    --hash=sha256:663fb62ad39200c7d903e973aa0aa693578613264d83796455cbf4cd172bd878 \
-    --hash=sha256:59a76c4142478a1952bba6f9a9ca4fc7b029afb619e8ffcf0d135ce37ea692da
+elasticsearch==5.1.0 \
+    --hash=sha256:ee95918958508fd34e533f78d0790e0e5ee03d0942b8988d34e6b6c4467edf66 \
+    --hash=sha256:a0bc4508764aa14c0cd15840a0ab38020b68e17d699274417986e3da0cee82fa
+elasticsearch-dsl==5.1.0 \
```
Comment (peterbe) [inline requirements.txt]: There is now a `5.2.0` of both `elasticsearch` and `elasticsearch-dsl` in pypi. Is now maybe a good time to push to that too?

**Comment by adngdb** [comment]
Rebased and updated to be compatible with Elasticsearch 5.3, cc @milescrabill.

**Comment by willkg** [comment]
This has conflicts. I'm going to see if I can address them now. Then I'll update the branch and we can squash and merge.

🎉
